mail_tracking/tests/test_mail_tracking.py

```diff
@@ -0,0 +1,706 @@
+        )
+        self.last_request = http.request
+        http.request = type(
+            "obj",
+            (object,),
+            {
+                "env": self.env,
+                "cr": self.env.cr,
+                "db": self.env.cr.dbname,
+                "endpoint": type("obj", (object,), {"routing": []}),
+                "httprequest": type(
+                    "obj",
+                    (object,),
+                    {"remote_addr": "123.123.123.123", "user_agent": FakeUserAgent()},
+                ),
+            },
+        )
+        for _ in http._generate_routing_rules(
+            ["mail", "mail_tracking"], nodb_only=False
+        ):
+            pass
+
+    def tearDown(self, *args, **kwargs):
+        http.request = self.last_request
+        return super(TestMailTracking, self).tearDown(*args, **kwargs)
+
+    def test_empty_email(self):
+        self.recipient.write({"email_bounced": True})
+        self.recipient.write({"email": False})
+        self.assertEqual(False, self.recipient.email)
+        self.assertEqual(False, self.recipient.email_bounced)
+        self.recipient.write({"email_bounced": True})
+        self.recipient.write({"email": ""})
+        self.assertEqual(False, self.recipient.email_bounced)
+        self.assertEqual(False, self.env["mail.tracking.email"].email_is_bounced(False))
+        self.assertEqual(
+            0.0, self.env["mail.tracking.email"].email_score_from_email(False)
+        )
+
+    def test_recipient_address_compute(self):
+        mail, tracking = self.mail_send(self.recipient.email)
+        tracking.write({"recipient": False})
+        self.assertEqual(False, tracking.recipient_address)
+
+    def test_message_post(self):
+        # This message will generate a notification for recipient
+        message = self.env["mail.message"].create(
+            {
+                "subject": "Message test",
+                "author_id": self.sender.id,
+                "email_from": self.sender.email,
+                "message_type": "comment",
+                "model": "res.partner",
+                "res_id": self.recipient.id,
+                "partner_ids": [Command.link(self.recipient.id)],
+                "body": "<p>This is a test message</p>",
+            }
+        )
+        if message.is_thread_message():
+            self.env[message.model].browse(message.res_id)._notify_thread(message)
+        # Search tracking created
+        tracking_email = self.env["mail.tracking.email"].search(
+            [
+                ("mail_message_id", "=", message.id),
+                ("partner_id", "=", self.recipient.id),
+            ]
+        )
+        # The tracking email must be sent
+        self.assertTrue(tracking_email)
+        self.assertEqual(tracking_email.state, "sent")
+        # message_dict read by web interface
+        message_dict = message.message_format()[0]
+        self.assertTrue(len(message_dict["history_partner_ids"]) > 0)
```

Comment:
`assertGreater(..., 0)`

But since length can't be negative and what you actually want to check it's set, the following would be enough I think:

`assertTrue(message_dict["history_partner_ids"])`